drivers: i2c: silabs: i2c driver for EFR series 2 devices

fimohame opened the pull request:
This PR introduces I2C driver implementation for Silicon Labs EFR Series 2 devices.

fimohame committed:
drivers: i2c: silabs: i2c driver for EFR series 2 devices

Added the i2c driver for EFR series 2 devices.

Signed-off-by: S Mohamed Fiaz <Fiaz.Mohamed@silabs.com>

Martinhoff-maker commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -0,0 +1,511 @@
+#include <zephyr/pm/device.h>
+#include <zephyr/pm/device_runtime.h>
+#include <zephyr/pm/policy.h>
+#include <sl_hal_i2c.h>
```
this include is already included in sl_i2c.c

Martinhoff-maker commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -0,0 +1,511 @@
+#include <zephyr/pm/policy.h>
+#include <sl_hal_i2c.h>
+#include <sl_i2c.h>
+#include <sli_i2c.h>
```
this one too

fimohame replied:
In zephyr driver we use sli_i2c_instance_t and functions

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -0,0 +1,511 @@
+	const struct i2c_silabs_dev_config *config = dev->config;
+	struct i2c_silabs_dev_data *data = dev->data;
+	sl_i2c_init_params_t init_params;
+	uint32_t baudrate;
```
addressed

fimohame committed:
drivers: i2c: silabs: add async combined write-read

Added support for combined write-read operations
with repeated start for asynchronous (async)
operations in the Silabs I2C driver for EFR
series 2 devices.

Signed-off-by: S Mohamed Fiaz <Fiaz.Mohamed@silabs.com>

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -0,0 +1,507 @@
+		data->dma_tx.dma_channel = dma_request_channel(data->dma_tx.dma_dev, NULL);
+		data->i2c_instance.dma_channel.dma_tx_channel = data->dma_tx.dma_channel;
+
+		if (data->dma_rx.dma_channel < 0 || data->dma_tx.dma_channel < 0) {
```
addressed

fimohame committed:
west.yml: Update hal_silabs revision

This commit updates the hal_silabs revision

Signed-off-by: S Mohamed Fiaz <Fiaz.Mohamed@silabs.com>

teburd commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -0,0 +1,529 @@
+{
+	struct i2c_silabs_dev_data *data = dev->data;
+
+	if (!k_sem_count_get(&data->bus_lock)) {
```
Hmm? What if you are interrupted between here and the k_sem_take(bus_lock, K_FOREVER) where another context does a transfer?

This isn't quite right. In the asynchronous case you should be doing k_sem_take(bus_lock, K_NO_WAIT) otherwise in the ISR context (which this API allows!) you will block, in the ISR, potentially forever.

fimohame replied:
addressed

fimohame commented on drivers/i2c/i2c_silabs.c:
```diff
@@ -0,0 +1,513 @@
+#include <zephyr/logging/log.h>
+#if defined(CONFIG_I2C_SILABS_DMA)
+#include <zephyr/drivers/dma.h>
+#endif
```
Acknowledged.
